File: Python/Problems/Problems/problem1563.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem1563(Problem):
    def test(self) -> bool:
        stoneValue = [6, 2, 3, 4, 5, 5]

        expected = 18

        result = self.stoneGameV(stoneValue)

        print(result)

        return result == expected

    def stoneGameV(self, stoneValue: List[int]) -> int:
        n = len(stoneValue)
        dp = [[-1 for _ in range(n)] for _ in range(n)]
        pref = [0]
        for i in range(n):
            pref.append(pref[i] + stoneValue[i])
        return self.gameDp(dp, pref, stoneValue, 0, n - 1)

    def gameDp(self, dp: List[List[int]], pref: List[int], arr: List[int], l: int, r: int) -> int:
        if l == r:
            return 0
        if dp[l][r]!= -1:
            return dp[l][r]
        total = pref[r + 1] - pref[l]
        pre = 0
        best = 0
        for i in range(l, r):
            pre += arr[i]
            suf = total - pre
            if pre > suf:
                best = max(best, suf + self.gameDp(dp, pref, arr, i + 1, r))
            elif pre < suf:
                best = max(best, pre + self.gameDp(dp, pref, arr, l, i))
            else:
                best = max(best, max(pre + self.gameDp(dp, pref, arr, l, i), suf + self.gameDp(dp, pref, arr, i + 1, r)))
        dp[l][r] = best
        return best

File: C++/Problems/Problems/problem_1563.cpp
#include "problem_1563.h"

bool problem_1563::test() {
    std::vector stoneValue { 6, 2, 3, 4, 5, 5 };

    int expected = 18;

    auto result = stoneGameV(stoneValue);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_1563::stoneGameV(std::vector<int> &stoneValue) {
    int n = stoneValue.size();
    std::vector dp(n, std::vector(n, -1));;
    std::vector pref(n + 1, 0);
    for (int i = 0; i < n; ++i)
        pref[i + 1] = pref[i] + stoneValue[i];
    return gameDp(dp, pref, stoneValue, 0, n - 1);
}

int problem_1563::gameDp(std::vector<std::vector<int>> &dp, std::vector<int> &pref, std::vector<int> &arr, int l, int r) {
    if (l == r)
        return 0;
    if (dp[l][r] != -1)
        return dp[l][r];
    int total = pref[r + 1] - pref[l];
    int pre = 0;
    int best = 0;
    for (int i = l; i < r; ++i)
    {
        pre += arr[i];
        int suf = total - pre;
        if (pre > suf)
            best = std::max(best, suf + gameDp(dp, pref, arr, i + 1, r));
        else if (pre < suf)
            best = std::max(best, pre + gameDp(dp, pref, arr, l, i));
        else
            best = std::max(best, std::max(pre + gameDp(dp, pref, arr, l, i), suf + gameDp(dp, pref, arr, i + 1, r)));
    }
    dp[l][r] = best;
    return best;
}
